Add detection of Mainboard firmware type( Marlin or Smoothieware), auto-detect LONG_FILENAME_HOST_SUPPORT, on-boardSD and serial_always_on option in config file & other bugfixes

guruathwal opened the pull request:
- Add detection of Mainboard firmware type( Marlin or Smoothieware)
Smoothieware does not support M115 Detailed reporting. Hence some settings need to be enabled or disabled in TFT for Smoothieware.
- Fix text display errors.
- Fix to prevent settings reset after config update.
- add Onboard SD card options for non-marlin firmware in the config file.
- add serial always On option in the config file.
- add missing Hotend count in the config file.
- Fix preheat temps. not updating in the config file.
- Fix fan max PWM update in the config file.
- Fix parameter settings not saving (#651).
- Remove all traces of EMERGENCY_PARSER.
- Fix Babystepping Z offset.
- Add auto-detect LONG_FILENAME_HOST_SUPPORT.
- Allow negative values in machine size (#698)
- Fix broken pause/ resume gcode

Resolves #668 , resolves #651, resolves #626 , resolves #698

guruathwal commented:
@digant73  Thanks for checking. Still, some of the replacements are missed. I will try to add these corrections as soon as possible.

digant73 commented:
> @digant73 Thanks for checking. Still, some of the replacements are missed. I will try to add these corrections as soon as possible.

no problem.
If you want to provide also PR #691 please consider the attached file (and not the one provided in the previous zip file).
That file fixes the bug on displaying 6 tool temperatures instead of the ones for the configured tools

guruathwal commented:
@digant73  already fixed it.. just pushing the commit. 👍

SonBroku commented:
So far, all the problems I've been having are fixed with your branch now @guruathwal :)

For sure this needs to be merged into the master branch.

extesy commented:
@bigtreetech can you merge please?

digant73 commented:
> Thank you sir!
> But I am somewhat confused right now, because I did not change _**config.ini**_ at all. But still all the settings which were set in _**Configuration.h**_, were applied to the _***.bin**_ file accordingly.
> What is _Configuration.h_ for, if _config.ini_ does the job? (Sorry for the digging. But I am just curious.)
> 
> As I mentioned before, I am compiling not for the BTT hardware, but for an MKS display,used in the _**Artillery Sidewinder X1 v4**_.
> 
> Could you please look into the attached files, which I got @delwinbest (**Who, by the way, did a great job!** **[here](https://github.com/delwinbest/BIGTREETECH-TouchScreenFirmware/tree/bibo2_touch)**) and customized it for my needs?
> 
> Because as also mentioned above, I set all the needed parameters in _**Configuration.h**_ and the _***.bin**_ file was compiled just as I wanted it to, despite some minor issues , I am wondering why the _**config.ini**_ was not considered for compilation?
> 
> [BTT-MKS.zip](https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/files/4646316/BTT-MKS.zip)
> 
> thank you in advance
> regards

I have also the Sidewinder X1 v4. The Configuration.h file is still present just to provide default values for the parameters defined in the config.ini file. So if a parameter is missing (e.g. commented out) in the config.ini file, it's default value will be used. The meaning for config.ini is to allow to set the parameters without the need to change the hard coded values in the Configuration.h file and rebuilding the .bin

guruathwal commented:
🤔 I wonder why I keep missing those... are you sure these are last??

digant73 commented:
>  I wonder why I keep missing those... are you sure these are last??
> 
> PS: Most of those were actually fixed but I don't know why those changes did not get saved.

Yes, those should be the last bug fixes (at least I found :-)) plus the possible 3 bugs that should be checked. Theese possible 3 bugs are not related to your changes. They were present before you mixed your PR. But it should be better to check them in order to fix all in one release

guruathwal commented:
@digant73 I checked the printgcodes are saved and loaded from SPI flash correctly. can you provide more detail about the bug you found?

digant73 commented:
> @digant73 I checked the printgcodes are saved and loaded from SPI flash correctly. can you provide more detail about the bug you found?

with your code (1 week ago) the cancel gcode was not executed. However let me check again with your last code. I will let you know

digant73 commented:
> > @digant73 I checked the printgcodes are saved and loaded from SPI flash correctly. can you provide more detail about the bug you found?
> 
> with your code (1 week ago) the cancel gcode was not executed. However let me check again with your last code. I will let you know

Ok I tested again the cancel gcode and it seems that the gcode creates some problems. I mean the gcode seems not providing the expected result (I used G28) and also after that gcode is submitted the system will start to work not properly. E.g. from Terminal menu I provided the same gcode and even in this case there is no expected result. It seems that Marlin is not processing the command and not providing any response.
If I restart the printer and from Terminal menu I provide the same gcode, the gcode is executed correctly and Marlin provides echo responses. So it seems that when gcode is executed on Printing.c the system become not stable. 

Below the parameter related to cancel gcode I configured on config.in:
cancel_gcode_enabled:1
cancel_gcode:G28 XY R10\n

Please use the same gcode

guruathwal commented:
@digant73 I tried the cancel gcode you provided and also changing it 3 time to make sure the changes code is being updated, and it was working. here is the screenshot of the serial comm I tried to test. 
![image](https://user-images.githubusercontent.com/22422544/82411649-c7b8cf80-9a8f-11ea-8553-ba7f403519bf.png)

Also, can you tell the total number of config keywords shown in the update progress?

digant73 commented:
> @digant73 I tried the cancel gcode you provided and also changing it 3 time to make sure the changes code is being updated, and it was working. here is the screenshot of the serial comm I tried to test.
> ![image](https://user-images.githubusercontent.com/22422544/82411649-c7b8cf80-9a8f-11ea-8553-ba7f403519bf.png)
> 
> Also, can you tell the total number of config keywords shown in the update progress?
> 
> @discip we are discussing the same.

91 entries successfully loaded from config.ini file (no error displayed)
In the abortPrinting function if I replace "printcodes->cancel_gcode" with the hard coded string "G28 XY R10\n" it works.
Furthermore if I pause a printing, when I resume it I got some beeps and messages displayed in the screen. The last displayed message is "v76 stored settings retrieved (635 bytes; crc 9439)". The printing is not resumed.
I'm working on MKS TFT28 (Artillery Sidewinder X1 printer).
I'm trying to display the "printcodes->cancel_gcode" when printing to see which value is loaded

guruathwal commented:
> 91 entries successfully loaded from config.ini file (no error displayed)
> In the abortPrinting function if I replace "printcodes->cancel_gcode" with the hard coded string "G28 XY R10\n" it works.
> Furthermore if I pause a printing, when I resume it I got some beeps and messages displayed in the screen. The last displayed message is "v76 stored settings retrieved (635 bytes; crc 9439)". The printing is not resumed.
> I'm working on MKS TFT28 (Artillery Sidewinder X1 printer).
> I'm trying to display the "printcodes->cancel_gcode" when printing to see which value is loaded

That message is related to **M501**. are you sure you did not sent it or added if any where?
Can you add this code in **Printing.c** after line no.187? see what info does is shows when you start a print
```c
 popupReminder((u8*)"gcode test",(u8*)printcodes->cancel_gcode);
```

digant73 commented:
> > 91 entries successfully loaded from config.ini file (no error displayed)
> > In the abortPrinting function if I replace "printcodes->cancel_gcode" with the hard coded string "G28 XY R10\n" it works.
> > Furthermore if I pause a printing, when I resume it I got some beeps and messages displayed in the screen. The last displayed message is "v76 stored settings retrieved (635 bytes; crc 9439)". The printing is not resumed.
> > I'm working on MKS TFT28 (Artillery Sidewinder X1 printer).
> > I'm trying to display the "printcodes->cancel_gcode" when printing to see which value is loaded
> 
> That message is related to **M501**. are you sure you did not send it or added it anywhere?
> Can you add this code in **Printing.c** after line no.187? see what info does is shows when you start a print
> 
> ```c
>  popupReminder((u8*)"gcode test",(u8*)printcodes->cancel_gcode);
> ```

Yes I didn't added any M501 when aborting a print. I'm using the same Printing.c file you committed yesterday with my last changes. However I will check that problem after.
I also printed the cancel gcode and its lenght. Its a zero lenght string.
I will integrate your last commit in my project and try again

digant73 commented:
> > > 91 entries successfully loaded from config.ini file (no error displayed)
> > > In the abortPrinting function if I replace "printcodes->cancel_gcode" with the hard coded string "G28 XY R10\n" it works.
> > > Furthermore if I pause a printing, when I resume it I got some beeps and messages displayed in the screen. The last displayed message is "v76 stored settings retrieved (635 bytes; crc 9439)". The printing is not resumed.
> > > I'm working on MKS TFT28 (Artillery Sidewinder X1 printer).
> > > I'm trying to display the "printcodes->cancel_gcode" when printing to see which value is loaded
> > 
> > 
> > That message is related to **M501**. are you sure you did not send it or added it anywhere?
> > Can you add this code in **Printing.c** after line no.187? see what info does is shows when you start a print
> > ```c
> >  popupReminder((u8*)"gcode test",(u8*)printcodes->cancel_gcode);
> > ```
> 
> Yes I didn't added any M501 when pausing a print. I'm using the same Printing.c file you committed yesterday with my last changes. However I will check that problem after.
> I also printed the cancel gcode and its lenght. Its a zero lenght string.
> I will integrate your last commit in my project and try again

I found that putting an empty string on the queue makes the system unstable. So mustStoreCmd and storeCmd function must be improved making a check with an empty string (avoid to put an empty string in the queue). With this fix, at least if printcodes->cancel_gcode is an empty string, as in my case, it will not make the system unstable

discip commented:
> > Thank you sir!
> > But I am somewhat confused right now, because I did not change _**config.ini**_ at all. But still all the settings which were set in _**Configuration.h**_, were applied to the _***.bin**_ file accordingly.
> > What is _Configuration.h_ for, if _config.ini_ does the job? (Sorry for the digging. But I am just curious.)
> > As I mentioned before, I am compiling not for the BTT hardware, but for an MKS display,used in the _**Artillery Sidewinder X1 v4**_.
> > Could you please look into the attached files, which I got @delwinbest (**Who, by the way, did a great job!** **[here](https://github.com/delwinbest/BIGTREETECH-TouchScreenFirmware/tree/bibo2_touch)**) and customized it for my needs?
> > Because as also mentioned above, I set all the needed parameters in _**Configuration.h**_ and the _***.bin**_ file was compiled just as I wanted it to, despite some minor issues , I am wondering why the _**config.ini**_ was not considered for compilation?
> > [BTT-MKS.zip](https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/files/4646316/BTT-MKS.zip)
> > thank you in advance
> > regards
> 
> I have also the Sidewinder X1 v4. The Configuration.h file is still present just to provide default values for the parameters defined in the config.ini file. So if a parameter is missing (e.g. commented out) in the config.ini file, it's default value will be used. The meaning for config.ini is to allow to set the parameters without the need to change the hard coded values in the Configuration.h file and rebuilding the .bin

@digant73
Good morning sir,
thank you for the clarification on that. In fact I did not understand all of it at that time. And in addition to that, I am not that familiar with the github environment yet, so I did not see your answer in time, because I thought my post was attached to the specific commit of @guruathwal only.

Since you are running the same hardware (I am not completely sure about that though.), I would love to get your code you are running your ASWX1 with (Marlin & TFT).
Is there any chance you would share it?

thank you in advance
kind regards

digant73 commented:
> > > Thank you sir!
> > > But I am somewhat confused right now, because I did not change _**config.ini**_ at all. But still all the settings which were set in _**Configuration.h**_, were applied to the _***.bin**_ file accordingly.
> > > What is _Configuration.h_ for, if _config.ini_ does the job? (Sorry for the digging. But I am just curious.)
> > > As I mentioned before, I am compiling not for the BTT hardware, but for an MKS display,used in the _**Artillery Sidewinder X1 v4**_.
> > > Could you please look into the attached files, which I got @delwinbest (**Who, by the way, did a great job!** **[here](https://github.com/delwinbest/BIGTREETECH-TouchScreenFirmware/tree/bibo2_touch)**) and customized it for my needs?
> > > Because as also mentioned above, I set all the needed parameters in _**Configuration.h**_ and the _***.bin**_ file was compiled just as I wanted it to, despite some minor issues , I am wondering why the _**config.ini**_ was not considered for compilation?
> > > [BTT-MKS.zip](https://github.com/guruathwal/BIGTREETECH-TouchScreenFirmware/files/4646316/BTT-MKS.zip)
> > > thank you in advance
> > > regards
> > 
> > 
> > I have also the Sidewinder X1 v4. The Configuration.h file is still present just to provide default values for the parameters defined in the config.ini file. So if a parameter is missing (e.g. commented out) in the config.ini file, it's default value will be used. The meaning for config.ini is to allow to set the parameters without the need to change the hard coded values in the Configuration.h file and rebuilding the .bin
> 
> @digant73
> Good afternoon sir,
> thank you for the clarification on that. In fact I did not understand all of it at that time. And in addition to that, I am not that familiar with the github environment yet, so I did not see your answer in time, because I thought my post was attached to the specific commit of @guruathwal only.
> 
> Since you are running the same hardware (I am not completely sure about that though.), I would love to get your code you are running your ASWX1 with (Marlin & TFT).
> Is there any chance you would share it?
> 
> thank you in advance
> kind regards

As Marlin firmware I'm using this https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/663#issuecomment-631381520. Below you find the zip file.
As TFT firmware I'm currently using the last committed version provided by gutuathwal. Move on His "Commit" tab and download his last commit you found at the end of that the page. However, as you can see it is still WIP (work in progress) so I suggest you to wait the bugs are solved and this PR is merged by BTT on their master branch

[ASWX1-FW-MOD-v1.2.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/4656458/ASWX1-FW-MOD-v1.2.zip)

digant73 commented:
> > > > 91 entries successfully loaded from config.ini file (no error displayed)
> > > > In the abortPrinting function if I replace "printcodes->cancel_gcode" with the hard coded string "G28 XY R10\n" it works.
> > > > Furthermore if I pause a printing, when I resume it I got some beeps and messages displayed in the screen. The last displayed message is "v76 stored settings retrieved (635 bytes; crc 9439)". The printing is not resumed.
> > > > I'm working on MKS TFT28 (Artillery Sidewinder X1 printer).
> > > > I'm trying to display the "printcodes->cancel_gcode" when printing to see which value is loaded
> > > 
> > > 
> > > That message is related to **M501**. are you sure you did not send it or added it anywhere?
> > > Can you add this code in **Printing.c** after line no.187? see what info does is shows when you start a print
> > > ```c
> > >  popupReminder((u8*)"gcode test",(u8*)printcodes->cancel_gcode);
> > > ```
> > 
> > 
> > Yes I didn't added any M501 when pausing a print. I'm using the same Printing.c file you committed yesterday with my last changes. However I will check that problem after.
> > I also printed the cancel gcode and its lenght. Its a zero lenght string.
> > I will integrate your last commit in my project and try again
> 
> I verified that putting an empty string on the queue makes the system unstable. So mustStoreCmd and storeCmd function must be improved making a check with an empty string (avoid to put an empty string in the queue). With this fix, at least if printcodes->cancel_gcode is an empty string, as in my case, it will not make the system unstable

Ok I found the problems:
1) Printing.c:
  replace

  uint8_t *data_p = (uint8_t *)printcodes;
  W25Qxx_ReadBuffer(data_p,PRINT_GCODES_ADDR,sizeof(printcodes));

  with

  W25Qxx_ReadBuffer((u8*)printcodes,PRINT_GCODES_ADDR,sizeof(PRINT_GCODES));

  The problem was of using "sizeof(printcodes)" instead of sizeof(PRINT_GCODES).
  With the above changes, the structues are acllocated as you made for custom gcodes on MachineSettings.c

  I suspect that you had a different Printing.c file. With that bug you read only 8 bytes on printcodes

2) With the above fix you are able to load start and end gcodes only. In order the cancel gcode is loaded you must put a new line after "cancel_gcode:xxxxxxxx" on the config.ini file. The config.ini is in unix format. Are you sure your code is able to parse unix format? At least it is better to provide after the "cancel_gcode:" a big comment asking to provide a new line after the last parameter defined in the config.ini file

guruathwal commented:
Hi,
I just changed the way print codes are sent.

guruathwal commented:
> Ok I found the problems:
> 
> 1. Printing.c:
> 
> replace
> 
> uint8_t *data_p = (uint8_t *)printcodes;
> W25Qxx_ReadBuffer(data_p,PRINT_GCODES_ADDR,sizeof(printcodes));
> 
> with
> 
> W25Qxx_ReadBuffer((u8*)printcodes,PRINT_GCODES_ADDR,sizeof(PRINT_GCODES));
> 
> The problem was due to the usage of "sizeof(printcodes)" instead of sizeof(PRINT_GCODES).
> With the above changes, the structues are allocated as you made for custom gcodes on MachineSettings.c
> 
> I suspect that you had a different Printing.c file. With that bug you loaded always only 8 bytes on printcodes. That bug is probably the reason of many people complaining of instability during printing. Access to unallocated memory was performed every time printcodes.start_gcode etc... were accessed
> 
> 1. With the above fix you are able to load start and end gcodes only. In order the cancel gcode is loaded you must put a new line after "cancel_gcode:" parameter" on the config.ini file. It is better to provide a big comment after "cancel_gcode:xxxx" asking to provide a new line after the last parameter defined in the config.ini file

Yes, I also sorted it out, and also changed the procedure to avoid the use of `malloc`.
I also added debug option in config.c to display the print codes read from the config file.

digant73 commented:
> > Ok I found the problems:
> > 
> > 1. Printing.c:
> > 
> > replace
> > uint8_t *data_p = (uint8_t *)printcodes;
> > W25Qxx_ReadBuffer(data_p,PRINT_GCODES_ADDR,sizeof(printcodes));
> > with
> > W25Qxx_ReadBuffer((u8*)printcodes,PRINT_GCODES_ADDR,sizeof(PRINT_GCODES));
> > The problem was due to the usage of "sizeof(printcodes)" instead of sizeof(PRINT_GCODES).
> > With the above changes, the structues are allocated as you made for custom gcodes on MachineSettings.c
> > I suspect that you had a different Printing.c file. With that bug you loaded always only 8 bytes on printcodes. That bug is probably the reason of many people complaining of instability during printing. Access to unallocated memory was performed every time printcodes.start_gcode etc... were accessed
> > 
> > 1. With the above fix you are able to load start and end gcodes only. In order the cancel gcode is loaded you must put a new line after "cancel_gcode:" parameter" on the config.ini file. It is better to provide a big comment after "cancel_gcode:xxxx" asking to provide a new line after the last parameter defined in the config.ini file
> 
> Yes, I also sorted it out, and also changed the procedure to avoid the use of `malloc`.
> I also added debug option in config.c to display the print codes read from the config file.
> Please remember to disconnect the TFT from the printer when using debug option as it will also dump the data to the serial port.

I saw the changes. Yes the problem is fixed. Remember to provide a comment or something like that at the end of the config.ini file to avoid the cancel_gcode is not loaded

guruathwal commented:
> I saw the changes. Yes the problem is fixed. Remember to provide a comment or something like that at the end of the config.ini file to avoid the cancel_gcode is not loaded

Yes, I am changing config.c to force parsing at the end of the file. ✌

digant73 commented:
> > I saw the changes. Yes the problem is fixed. Remember to provide a comment or something like that at the end of the config.ini file to avoid the cancel_gcode is not loaded
> 
> Yes, I am changing config.c to force parsing at the end of the file.

still not working even with the change.

guruathwal commented:
Yes, it was parsing before converting the "\n" to a new line char. fixed it.

digant73 commented:
> Yes, it was parsing before converting the "\n" to a new line char. fixed it.
> 
> Tested both with and without a line ending.

hmm it seems not working yet. I have configured 88 parameters (so #88 is the cancel gcode).
If a new line is present after the last parameter (cancel gcode) all 88 parameters are loaded and everything wors as expected
If no new line is present after the last parameter (cancel gcode) only 87 parameters are loaded and aborting the print I also get the error "Unknown command: "E0""
Attached my config.ini file with no new line at the end. Please try to use it.
By default I disable all 3 printing gcode so you must first enable at least the cancel gcode from the "features menu" before printing
It seems to me that the error I get is related to the end gcode (E0 is present on it)

[config.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/4658024/config.zip)

guruathwal commented:
Hi,
found it. I forgot to add terminate character at the end of the line in config.c. So a part of end_gcode ended up in cancel_gcode.

Rsoft1976 commented on TFT/src/User/SanityCheck.h:
```diff
@@ -14,9 +14,9 @@
+#ifdef ST7920_SPI
     #ifdef CLEAN_MODE_SWITCHING_SUPPORT
-    #error "CLEAN_MODE_SWITCHING_SUPPORT can only be enabled for TFT controllers which support ST7920 Emulator/Marlin Mode. Disable CLEAN_MODE_SWITCHING_SUPPORT in Configuration.h"
+    #error "CLEAN_MODE_SWITCHING_SUPPORT is now SERIAL_ALWAYS_ON. Please update your configuration.
```
#error "CLEAN_MODE_SWITCHING_SUPPORT is now SERIAL_ALWAYS_ON. Please update your configuration.h"

digant73 commented:
> Hi,
> found it. I forgot to add terminate character at the end of the line in config.c. So a part of end_gcode ended up in cancel_gcode.

yes, it seems ok now. I see just a difference in case there is no new line after the last parameter. In that case it is shown 87 instead of 88 parameters loaded. However it seems all 88 parameters were correctly loaded.

PS: do not forget to comment out the line

#define CONFIG_DEBUG  // To be used only when calling 'getConfigFromFile()' after boot process

on Config.c. It is currently not commented

discip commented:
@guruathwal & @digant73
I just wanted to thank you for being persistent.
Thank you for your tireless work on making this finally somewhat close to perfect. 👍 👍 👍

digant73 commented:
> > Hi,
> > found it. I forgot to add terminate character at the end of the line in config.c. So a part of end_gcode ended up in cancel_gcode.
> 
> yes, it seems ok now. I see just a difference in case there is no new line after the last parameter. In that case it is shown 87 instead of 88 parameters loaded. However it seems all 88 parameters were correctly loaded.
> 
> PS: do not forget to comment out the line
> 
> #define CONFIG_DEBUG // To be used only when calling 'getConfigFromFile()' after boot process
> 
> on Config.c. It is currently not commented
> 
> Now the only problem is still present on Printing menu is on the resume print. When I resume a print, the nozzle moves correclty to the point where it was previous to the pause but after few seconds the steppers are disabled and I got one or more errors (I cannot say how many because they are displayed too fast on the display). The last one I see is related to M501 (it was executed M501).
> I don't know if the problem could be related to some echos provided by Marlin not correctly handled by TFT.
> However, after I got the error the system is not stable (temperatures are no longer updated, gcode from Terminal are not processed etc...) as it was when and empty string was put on the command queue. I need to restart the TFT

Just an update about the problem on pause/resume. I installed version 0.26 (so the official release from BTT before your integration) and resume works fine. Then I installed version 0.26.1 (so the official release from BTT with your integration) and resume is not working. So the problem is present from your first release (not from this PR)

guruathwal commented:
> Just an update about the problem on pause/resume. I installed version 0.26 (so the official release from BTT before your integration) and resume works fine. Then I installed version 0.26.1 (so the official release from BTT with your integration) and resume is not working. So the problem is present from your first release (not from this PR). Just trying to discover the problem

I tried to track all the changes from 0.26 till now but I can't find any significant change which may change the behaviour. To pinpoint the broken commit can you try with this commit: https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/a60960cabae40b65118c8cdd2e110fc619424484. It's just before adding the config file update.
Also, can you share the gcode file you are using to test so that I can replicate the error?

digant73 commented:
> > Just an update about the problem on pause/resume. I installed version 0.26 (so the official release from BTT before your integration) and resume works fine. Then I installed version 0.26.1 (so the official release from BTT with your integration) and resume is not working. So the problem is present from your first release (not from this PR). Just trying to discover the problem
> 
> I tried to track all the changes from 0.26 till now but I can't find any significant change which may change the behaviour. To pinpoint the broken commit can you try with this commit: [a60960c](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/a60960cabae40b65118c8cdd2e110fc619424484). It's just before adding the config file update.
> Also, can you share the gcode file you are using to test so that I can replicate the error?

ok, I will try with that commit and I let you know.
Attached the gcode file (however I also tried with other files getting same results).
Clearly differences with BTT release 0.26 is that your release moves the nozzle quicly on the right (on pause) while BTT on the left at lower speed. The same on resume, your release moves the nozzle quickly. Furthermore, just few seconds the nozzle is on resume position, the steppers are disabled and I get the errors

[Cap.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/4662702/Cap.zip)

guruathwal commented:
@digant73  added some changes to pause resume. can you check it?

digant73 commented:
> @digant73 added some changes to pause resume. can you check it?

should I 

> @digant73 added some changes to pause resume. can you check it?

Yes that was the problem. Now resume is working fine.
I previously checked also the commit you suggested and it was OK (there were just few changes compared to BTT release). You are probably checking all printf for all data types. That creates a lot of different behaviours in different HW.

Hukuma1 commented:
Love not knowing anything about coding and reading your guys' progress on squashing the bugs. Thank you! :)

digant73 commented:
thanks for the changes. They are better than my suggestions (you changed the % specifier and used promotion).
Just a possible change on Numpad.c.
In the instruction
 
my_sprintf(tempstr, "%d", old_val);

old_val has type u32 but you used %d. Maybe %u was the desired specifier

guruathwal commented:
@digant73 
The `my_sprintf` / `my_vsprintf` does not have the `%u` specifier. You can see below from line no. 94.
It is converting all input values of any size (signed or unsigned) to a signed int with `my_va_arg(p_next_arg, int);` at line no. 102.
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/56f904000f16f972a5a5ed3b5bfb121fac739023/TFT/src/User/my_misc.c#L67-L166

digant73 commented:
> @digant73
> The `my_sprintf` / `my_vsprintf` does not have the `%u` specifier. You can see below from line no. 94.
> It is converting all input values of any size (signed or unsigned) to a signed int with `my_va_arg(p_next_arg, int);` at line no. 102.
> https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/56f904000f16f972a5a5ed3b5bfb121fac739023/TFT/src/User/my_misc.c#L67-L166

right. The chinese comment in the code helped me a lot to understand!
Is there any list of possible reported bugs you are checking?

carl1961 commented:
google helps

Express integers in character form
For hexadecimal, 10 will be represented as A
Put integer + '0' = ASCII code corresponding to integer
So the result in str is upside down, flip it
Decimal integer
Hexadecimal integer
Floating point
String

On Fri, May 22, 2020 at 6:22 AM digant73 <notifications@github.com> wrote:

> @digant73 <https://github.com/digant73>
> The my_sprintf / my_vsprintf does not have the %u specifier. You can see
> below from line no. 94.
> It is converting all input values of any size (signed or unsigned) to a
> signed int with my_va_arg(p_next_arg, int); at line no. 102.
>
> https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/56f904000f16f972a5a5ed3b5bfb121fac739023/TFT/src/User/my_misc.c#L67-L166
>
> right. The chinese comment in the code helped me a lot to understand!
> Is there any list of possible reported bugs you are checking?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/663#issuecomment-632641394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXKJNPUAGHE2M56DNGZ4SLRSZOAZANCNFSM4M367U5A>
> .
>

digant73 commented:
> google helps , but seem the English is before the Chinese comments (I saw this after I posted LOL) Express integers in character form For hexadecimal, 10 will be represented as A Put integer + '0' = ASCII code corresponding to integer So the result in str is upside down, flip it Decimal integer Hexadecimal integer Floating point String
> […](#)
> On Fri, May 22, 2020 at 6:22 AM digant73 ***@***.***> wrote: @digant73 <https://github.com/digant73> The my_sprintf / my_vsprintf does not have the %u specifier. You can see below from line no. 94. It is converting all input values of any size (signed or unsigned) to a signed int with my_va_arg(p_next_arg, int); at line no. 102. https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/56f904000f16f972a5a5ed3b5bfb121fac739023/TFT/src/User/my_misc.c#L67-L166 right. The chinese comment in the code helped me a lot to understand! Is there any list of possible reported bugs you are checking? — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#663 (comment)](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/663#issuecomment-632641394)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAXKJNPUAGHE2M56DNGZ4SLRSZOAZANCNFSM4M367U5A> .

I was joking about the comment in chinese

digant73 commented:
@guruathwal According to the bug reporting no save on EEPROM was made on exiting from ParameterSettings menu, I honestly don't like that a permanent change is made without the user is prompted with a question asking if he wants to proceed to make permanent the changes. Furthermore (as it was in my case) someone could change the settings just for tuning the printer and testing the results (e.g. the steps-per-unit change) before he wants really make them permanent.
I had already made (for other usage) two dialog box handling the eeprom save and reset and calling them when needed (on my project I added new menus and fixed existing ones such as BabyStep, ProbeOffset, PreheatMenu, UnifiedHeat).
The changes I made are on top of your PR, so please let me know if you agree to change the current logic to always save on eeprom without prompting any message. I will provide you the changes on top of your PR.

Just about the bugs and improvements (not related to your PR) I found and fixed in my project that could be integrated in your PR:
1) babystep menu:
  1) the range limit of -5.00 to 5.00 could be not reached changing the unit value. E.g. if step is currently set to 4.10 and you change the unit value to 1.00 you cannot increment the step (it will remain 4.10 instead of becoming 5.00)
  2) The most important bug is that if you have set step to a value > 3.00 or < 3.00 the reset button will not work properly. The maximun value you can increase or decreese to Z-offset is 3.00. E.g. starting from an initial value of 0.00 for step and 0.00 for Z-offset, if you set step to 5.00 (so Z-offset will be increased to 5.00) then when you press on the reset button, step will become 0.00 but z-offset will be 2.00 (a maximum of 3.00 was decreased from z-offset on I single gcode invocation). I found that bug because I changed the babystep menu displaying also the z-offset. 
  Both bugs were fixed on my project
3) ProbeOffset menu:
  1) same bug on range limits as in babystep menu. Here the range limit is - 20 to +20 but you normally cannot reach that limits
  2) resetting the offset must also reset babystep. This is not done instead
  Both bugs were fixed on my project
4) Preheat menu:
  1) No button to stop pheheat is present.
  I added a stop button (using existing stop icon so no need to import new icon, label etc...) allowing to stop both, bed or nozzle preheat
5) UnifiedHeat menu:
  1) Cooldown button stops only the heaters and not also the fans (Fan menu is also present on that menu). It should stop both heaters and fans
  Now, when cooldown button is pressed both heaters and fans are stopped
All of those fixes/improvements are already on top of your PR, fully tested since several weeks.

Please let me know if you would consider to integrate that fixes/enhancements in your PR. I will eventually open a PR for that but this will take longer time to be eventually accepted and integrated in the master branch.

radek8 commented:
Can you please add the fixes for using the coder that I suggested in
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/690

discip commented:
@digant73
Good morning,
just as you might have expected, I am very curious about those fixes/ enhancements you posted about above, so is there any chance you would provide them?

thank you in advance
regards

guruathwal commented:
@digant73  I do not have any list of known bugs. I try to fix problems I find while using these devices, or a reported issue on Github or if anyone asks to solve a problem.

> @guruathwal According to the bug reporting no save on EEPROM was made on exiting from ParameterSettings menu, I honestly don't like that a permanent change is made without the user is prompted with a question asking if he wants to proceed to make permanent the changes. Furthermore (as it was in my case) someone could change the settings just for tuning the printer and testing the results (e.g. the steps-per-unit change) before he wants really make them permanent.

That's a nice idea. I never had this as I prefer saving the settings in case I forgot to save after printing or before shutting down / reset.

> 1. babystep menu:
>    *) the range limit of -5.00 to 5.00 could be not reached changing the unit value. E.g. if step is currently set to 4.10 and you change the unit value to 1.00 you cannot increment the step (it will remain 4.10 instead of becoming 5.00)
>    *) The most important bug is that if you have set step to a value > 3.00 or < -3.00 the reset button will not work properly. The maximun value you can increase or decreese to Z-offset is 3.00. E.g. starting from an initial value of 0.00 for step and 0.00 for Z-offset, if you set step to 5.00 (so Z-offset will be increased to 5.00) then when you press on the reset button, step will become 0.00 but z-offset will be 2.00 (a maximum of 3.00 was decreased from z-offset on I single gcode invocation). I found that bug because I changed the babystep menu displaying also the z-offset.
>    Both bugs were fixed on my project
> 2. ProbeOffset menu:
>    *) same bug on range limits as in babystep menu. Here the range limit is - 20 to +20 but you normally cannot reach that limits
>    *) resetting the offset must also reset babystep. This is not done instead
>    Both bugs were fixed on my project

In the baby-stepping menu, while printing, the max range I had to set is +-1mm, and I never used the probe offset menu. So I do not know much about this bug. If you can share the details and fix, it will be great. 
Although I doubt if while printing baby-stepping more than +-2mm is needed? if more than this is needed then will it not be a bad probe offset value then?

> 3. Preheat menu:
>    *) No button to stop pheheat is present.
>    I added a stop button (using existing stop icon so no need to import new icon, label etc...) allowing to stop both, bed or nozzle preheat

I am integrating the preheat options in the heating menu, so there will be a stop button in heat menu.

> 4. UnifiedHeat menu:
>    *) Cooldown button stops only the heaters and not also the fans (both Heat and Fan menus are present on that menu so the cooldown must be applied to both). It should stop both heaters and fans
>    Now, when cooldown button is pressed both heaters and fans are stopped

I did not add fan stop command as few people suggested to keep fans ON to speed the cooling process. Maybe it will be better to add cooldown command to the config file.


> Please let me know if you would consider to integrate that fixes/enhancements in your PR. I will eventually open a PR for that but this will take longer time to be eventually accepted and integrated in the master branch.

you can share the details or you can open a PR to my Mainstream branch, then it will be added to this PR after merging.

GregSKR commented:
@guruathwal Thanks for your work! 👍 
Looks like some of the changes you made in the code are reverted since BTT commit [46969db](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/663/commits/46969db22d6f706703884a76e7447983fe88a644)...
eg. for heat.c the use of 'HEATER_COUNT' has gone 
Same for Config.ini (cleaning code) & Printing.c (lines 742-745 added are not yours)...

digant73 commented:
> @digant73 I do not have any list of known bugs. I try to fix problems I find while using these devices, or a reported issue on Github or if anyone asks to solve a problem.
> 
> > @guruathwal According to the bug reporting no save on EEPROM was made on exiting from ParameterSettings menu, I honestly don't like that a permanent change is made without the user is prompted with a question asking if he wants to proceed to make permanent the changes. Furthermore (as it was in my case) someone could change the settings just for tuning the printer and testing the results (e.g. the steps-per-unit change) before he wants really make them permanent.
> 
> That's a nice idea. I never had this as I prefer saving the settings in case I forgot to save after printing or before shutting down / reset.

Good. I added the two functions "menuEepromSave" and "menuEepromReset" on MachineSettings.c and MachineSettings.h, so they can be called by any other file in the project.. Inside these functions a check if EEPROM is enabled or not is performed before eventually showing the popup screen. In case EEPROM is not enabled on Marlin, no popup is showed at all and the control is immediately returned to the caller (so it is absolutely transparent for the end user).
I also replaced any invocation of "M500" in the project with "menuEeepromSave". The affected files were "leveling.c" and the mentioned "Parametersettings.c". 3 new labels were added for showing the proper text on the popup dialog box. I provided only english translation for all the locales so in the future a proper translation could be done for any locale.
A small improvement was also applied on MachineSettings.c and Parametersettings.c. The "M503" command was moved from "menuMachineSettings" to "menuParameterSettings, so now the update of the parameters is made internally on menuParameterSettings before the parameters are displayed on the GUI.

> 
> > 1. babystep menu:
> >    *) the range limit of -5.00 to 5.00 could be not reached changing the unit value. E.g. if step is currently set to 4.10 and you change the unit value to 1.00 you cannot increment the step (it will remain 4.10 instead of becoming 5.00)
> >    *) The most important bug is that if you have set step to a value > 3.00 or < -3.00 the reset button will not work properly. The maximun value you can increase or decreese to Z-offset is 3.00. E.g. starting from an initial value of 0.00 for step and 0.00 for Z-offset, if you set step to 5.00 (so Z-offset will be increased to 5.00) then when you press on the reset button, step will become 0.00 but z-offset will be 2.00 (a maximum of 3.00 was decreased from z-offset on I single gcode invocation). I found that bug because I changed the babystep menu displaying also the z-offset.
> >    Both bugs were fixed on my project
> > 2. ProbeOffset menu:
> >    *) same bug on range limits as in babystep menu. Here the range limit is - 20 to +20 but you normally cannot reach that limits
> >    *) resetting the offset must also reset babystep. This is not done instead
> >    Both bugs were fixed on my project
> 
> In the baby-stepping menu, while printing, the max range I had to set is +-1mm, and I never used the probe offset menu. So I do not know much about this bug. If you can share the details and fix, it will be great.
> Although I doubt if while printing baby-stepping more than +-2mm is needed? if more than this is needed then will it not be a bad probe offset value then?
> 

Yes +-1.00 is the max unit you can use to increase or decrease the step value. But, for example, if you set the unit value to 0.1 and increase the step value to 4.4. After that you set the unit to 1 and try to increase the step value, the step value remains 4.4 instead of reaching the limit 5.0. That problem is probably present also on the heating menus but here the problem was pretty ease to be found due to the small range of admitted values.
I fixed this problem on both "BabyStep.c" and "ProbeOffset.c". In order to not write much code on the "switch" statement I preferred to organize the the code in functions that are then called on each "case" of the "switch" statement (e.g a decOffset function is called when pressing on dec button etc...).
About the possibility of setting a step value > 2 I agree with you (it means that the levelling is totally wrong and it should be performed again). However -5 +5 is the range provided by BTT (i didn't change it).
In order to guarantee that the reset of a step value > 3 is properly handled, I call the M290 command with a maximun of 1.00 unit. E.g. if step value is 5.00, M290 is called 5 times. E.g. if step value was 4.40, M290 is called 4 times with a 1.00 unit value and 1 time with 0.40 unit value.
Another bug solved on ProbeOffset is that every time you change a Z offset value, the babystep value is now reset to 0, If not reset, the set babystep value will not be aligned with the new z-offset. E.g., with the fix, from BabyStep menu you set baby step value to 5.00 and Z-offset is also 5.00. Then from ProbeOffset menu you change Z-offset to 3.00. If you move again on the BabyStep menu you will see that the babystep value is now reset to 0.00 while Z-offset is the expect value 3.00.
With the bug (baby step not reset), you will have baby step set to 5.00 and Z offset set to 3.00. If you click on reset button you will have 0.00 as baby step value and -2.00 as Z offset (you will probably scratch the bed if you move the extruder).

> > 1. Preheat menu:
> >    *) No button to stop pheheat is present.
> >    I added a stop button (using existing stop icon so no need to import new icon, label etc...) allowing to stop both, bed or nozzle preheat
> 
> I am integrating the preheat options in the heating menu, so there will be a stop button in heat menu.
> 

Ok, However attached you can also find my improvements for Fan.c and PreheatMenu.c.
On Fan.c, I replace the next fan selection button (useless because we have only one Fan selectable) with a more usefull half speed button. The Half speed icon is already present in the BTT project so no need of new icon was needed.
On PreheatMenu, The stop button is applied on the selected tool (both, nozzle or bed) as it happens for the abs, pla and petg buttons. An easy way to rollback a previously pressed preheat button.
Feel free to provide those improvements on a next PR or to skip them. For me they were pretty usefull.

> > 1. UnifiedHeat menu:
> >    *) Cooldown button stops only the heaters and not also the fans (both Heat and Fan menus are present on that menu so the cooldown must be applied to both). It should stop both heaters and fans
> >    Now, when cooldown button is pressed both heaters and fans are stopped
> 
> I did not add fan stop command as few people suggested to keep fans ON to speed the cooling process. Maybe it will be better to add cooldown command to the config file.
> 
> > Please let me know if you would consider to integrate that fixes/enhancements in your PR. I will eventually open a PR for that but this will take longer time to be eventually accepted and integrated in the master branch.
> 
> you can share the details or you can open a PR to my Mainstream branch, then it will be added to this PR after merging.


[PR_changes.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/4678005/PR_changes.zip)

digant73 commented:
> @digant73
> Good morning,
> just as you might have expected, I am very curious about those fixes/ enhancements you posted about above, so is there any chance you would provide them?
> 
> thank you in advance
> regards

zip file attached on my last reply to guruathwal

discip commented:
@digant73 
Thank you a ton! 😁
regards

SeanLMH commented:
Hi, I've tried to update my TFT35 with this but it still doesn't seem to detect the mainboard firmware type (so I'm getting cold extrusions when slicing with prusaslicer), is there something special I need to do with my Marlin config? I'm using the TFT35 V3 E3 with and SKR Mini E3 V1.2
